### PR TITLE
Add support c language

### PR DESCRIPTION
**Pull request by Mephistophiles**

### CODE REVIEW

**Comment by alexpovel** [comment]
This looks fantastic, appreciate the work! I’ll take a closer look soon but on first glance it looks great.

Two notes:

- CI might complain about unsorted imports. If you install a nightly Cargo toolchain and run cargo formatting (on nightly) it should sort itself out, no manual work required. The pre-commit config has the command for that. 
- let’s see about bumping dependencies, that has not been without pain in the past. Ideally this PR implements C support on the current tree-sitter version, then we bump versions separately (PRs for that exist already, #112 ). Does that work for you or do the new versions bring changes you need?

**Comment by Mephistophiles** [comment]
Hi! Thanks for the review!

> CI might complain about unsorted imports. If you install a nightly Cargo toolchain and run cargo formatting (on nightly) it should sort itself out, no manual work required. The pre-commit config has the command for that.
Hmm, I have the same formatting on cargo-fmt 1.83. I'll look a little later to see why it doesn't format for me.
> let’s see about bumping dependencies, that has not been without pain in the past
I'd like to use the latest version of tree-sitter-c, but it requires tree-sitter 0.23.
I'm still testing this parser, so feel free to merge PR with bump tree-sitter version. I'll just rebase my branch later and resolve the conflicts.

**Comment by alexpovel** [comment]
I pushed a change ( e847d3702e61a7c903abb970e36e3757d026ee38 ) that might address the CI failure at https://github.com/alexpovel/srgn/actions/runs/10971586945/job/30467773900?pr=116 in this PR. But no idea what's wrong there...

I'm looking into bumping tree-sitter versions. Not sure if https://github.com/tree-sitter/tree-sitter/pull/3069 introduced changes I need to respect here. But nothing for you to worry about.

**Comment by Mephistophiles** [comment]
Hi! I rebase my branch. I hope CI won't crash now =).
Can you tell me what other things should be done when adding support for a new language? (Should I add any additional tests, documentation?)

File: src/scoping/langs/c.rs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    fn neg_query(&self) -> Option<&TSQuery> {
+        self.negative_query.as_ref()
+    }
+}
+
+impl Find for C {
+    fn extensions(&self) -> &'static [&'static str] {
+        &["c"]
```
Comment (Mephistophiles) [inline src/scoping/langs/c.rs:124]:
```suggestion
        &["c", "h"]
```

File: src/main.rs [hunk 2]
```diff
@@ -1238,6 +1243,18 @@ mod cli {
         pub typescript: Option<TypeScriptScope>,
     }
 
+    #[derive(Parser, Debug, Clone)]
+    #[group(required = false, multiple = false)]
+    pub struct CScope {
+        /// Scope C code using a prepared query.
+        #[arg(long, env, verbatim_doc_comment, visible_alias = "c")]
```
Comment (alexpovel) [inline src/main.rs:1250]: You can leave off the `visible_alias`, it's identical to the normal command. Doesn't get shorter than `c` 😄

File: src/scoping/langs/c.rs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        self.negative_query.as_ref()
+    }
+}
+
+impl Find for C {
+    fn extensions(&self) -> &'static [&'static str] {
+        &["c", "h"]
+    }
+
+    fn is_path_invalid(&self, _: &Path) -> bool {
```
Comment (alexpovel) [inline src/scoping/langs/c.rs:127]: You can remove `is_path_invalid` altogether: it has a default implementation which does the exact same:

https://github.com/alexpovel/srgn/blob/45b64556df1abb983bd067112dacdbb645dd9b47/src/find/mod.rs#L31

File: README.md
Comment (alexpovel) [inline README.md:1]: If you like, you can add an example of what users can do with your new C functionality to the README, in this section:

https://github.com/alexpovel/srgn/blob/45b64556df1abb983bd067112dacdbb645dd9b47/README.md?plain=1#L776

There are some existing examples of what one might want to do with srgn.

This is optional though.

**Review by alexpovel — COMMENTED** [review]
@Mephistophiles , I'd be interested in your feedback. How was working on this for you? Adding support for a new language is not documented. Were there things you stumbled over (documentation, automation, ...)?

File: README.md
Comment (alexpovel) [reply README.md:1]: If there's other sections (next to "Prepared queries" showcases) of the README where you'd like to showcase your new work you can also do that, if you want.

File: src/main.rs (hunk 2 above)
Comment (Mephistophiles) [reply src/main.rs]: Oh, copypaste =) Deleted, thanks!

**Comment by Mephistophiles** [comment]
@alexpovel, When adding support for the C language, I relied on the comparison of how support was implemented for other languages. So, I didn't encounter any particular problems.
I really liked that readme was also tested, this allowed me to pay attention to the fact that readme should also be updated =).

Regarding automation with `pre-commit`: I have NixOS, so `commitizen` and `markdown-lint` did not work well for me due to the peculiarities of nixos (linker is located in completely different paths, different from the standard `/lib` `/usr/lib`), but this is just a peculiarity of my environment.

Regarding tests: do you think it is possible to automate the launch of tests for each prepared query language? This would allow updating test files (base files) in time. And another small question for the insta snapshot engine (I haven't worked with it much), is it possible to somehow detect old snapshot files that are no longer used? It happened to me that when renaming a prepared query, old files remained in staged, only rm -rf base.c_* helped =).

**Comment by Mephistophiles** [comment]
Oh, `Main / Execute release chores (pull_request)` still failing =( The crash couldn't be related to my commits?

**Review by alexpovel — APPROVED** [review]

**Comment by alexpovel** [comment]
Thanks for your feedback! Regarding your points:

- I run Nix (but on macOS). Perhaps one day I'll put a `flake.nix` into the repo to have everything truly fixed. I've done that in [simpler cases](https://github.com/alexpovel/rust-testing-explored/blob/aa62bc75ab30bb32a9959bf0f55cd30fb6f8284d/flake.nix) in the past and it works fine.
- The creation of test cases for https://github.com/alexpovel/srgn/blob/38cdec2541ab490b09aa2a8263525d34e20f83fd/tests/langs/mod.rs#L722 could surely be automated. But not sure it's worth it: it would be macro hell, and adding support for a new language/new test cases is quite rare. The repo has test coverage to see if cases were missed.
- You can run something like `cargo insta test --unreferenced=delete` to remove old snapshots. HOWEVER, that might not work with platform-dependent snapshots such as https://github.com/alexpovel/srgn/blob/38cdec2541ab490b09aa2a8263525d34e20f83fd/tests/snapshots/cli__tests__go-search-files-windows.snap , which are a stupid hack I introduced.
- The workflow `Main / Execute release chores (pull_request)` doesn't matter in PRs, all good :)

**Pull request merged**